File: 518. Coin Change 2.py
# // Time Complexity : O(n)
# // Space Complexity : O(n)
# // Did this code successfully run on Leetcode : Yes
# // Any problem you faced while coding this : No


# // Your code here along with comments explaining your approach
# Find the sub problem
# Formulate the dp matrix
# return the last element

class Solution:
    def change(self, amount: int, coins: List[int]) -> int:
        
        dp = [[0 for i in range(amount+1)] for j in range(len(coins)+1)]
        
        for i in range(len(dp)):
            dp[i][0] = 1
            
        for i in range(1,len(dp)):
            for j in range(0,len(dp[0])):
                if j < coins[i-1]:
                    dp[i][j] = dp[i-1][j]
                else:
                    dp[i][j] = dp[i-1][j] + dp[i][j-coins[i-1]]
        return dp[-1][-1]
            

File: _518_CoinChange2.java
// space complexity - o(n)
    	//time - o(n)
      //Ran on leeetcode successfully : Yes
      // Problem faced  : No
    	//Approach : linear memuization

class Solution {
    public int change(int amount, int[] coins) {

        if(amount == 0 ) return 1;
        if(coins == null || coins.length == 0 ) return 0;

        int[][] matrix  = new int[coins.length+1][amount+1];

        for(int i = 0; i<matrix.length; i++){
            for(int j = 0; j<matrix[0].length; j++){

                if(i == 0 || j == 0 ) {
                   matrix[i][j] = 0;
                }
                else if( j < coins[i-1]){
                    matrix[i][j] = matrix[i-1][j];
                }
                else if( j == coins[i-1]){
                    matrix[i][j] = 1 + matrix[i-1][j];
                }
                else if( j >= coins[i-1]){
                    int excludeCurrCoin = matrix[i-1][j];
                    int includeCurrCoin = matrix[i][j -coins[i-1]];
                    matrix[i][j] = excludeCurrCoin + includeCurrCoin;
                }
            }
        }

        return matrix[coins.length][amount];

    }
}
